### PR TITLE
Rewrite subplots_demo

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

This is a complete rewrite of https://matplotlib.org/gallery/subplots_axes_and_figures/subplots_demo.html

Improvements:
- Add links to the used API functions
- Structure the topics with headings
- Bring the topics in a logical order
- Add a short explanatory text to each example

### CODE REVIEW

File: examples/subplots_axes_and_figures/subplots_demo.py [hunk 1]
```diff
@@ -1,103 +1,192 @@
...
+
+# Hide x labels and tick labels for all but bottom plot.
+for ax in axs:
+    ax.label_outer()
+
+###############################################################################
+# Apart from ``True`` and ``False``, both *sharex* and *sharey* accept the
+# values 'row' and 'col' to share the values only per row or column.
+
+fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2,
```
Comment (anntzer) [inline examples/subplots_axes_and_figures/subplots_demo.py]: In general I don't like very short lines all squeezed to the right, I'd suggest e.g.
```
fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(
    2, 3, sharex='col', sharey='row',
    gridspec_kw={...})
```

File: examples/subplots_axes_and_figures/subplots_demo.py [hunk 2]
```diff
@@ -1,103 +1,192 @@
...
+# By default, each Axes is scaled individually. Thus, if the ranges are
+# different the tick values of the subplots do not align.
+
+fig, (ax1, ax2) = plt.subplots(2)
+fig.suptitle('Axes values are scaled individually by default')
+ax1.plot(x, y)
+ax2.plot(x + 1, -y)
+
+###############################################################################
+#
```
Comment (anntzer) [inline examples/subplots_axes_and_figures/subplots_demo.py]: extraneous empty line?

**Comment by NelleV** [comment]
I'll merge once CI passes.

File: examples/subplots_axes_and_figures/subplots_demo.py [hunk 3]
```diff
@@ -1,103 +1,190 @@
...
-axarr[0, 1].set_title('Axis [0,1]')
-axarr[1, 0].plot(x, y ** 2)
-axarr[1, 0].set_title('Axis [1,0]')
-axarr[1, 1].scatter(x, y ** 2)
-axarr[1, 1].set_title('Axis [1,1]')
-for ax in axarr.flat:
-    ax.set(xlabel='x-label', ylabel='y-label')
-# Hide x labels and tick labels for top plots and y ticks for right plots.
-for ax in axarr.flat:
+for ax in axs:
```
Comment (NelleV) [inline examples/subplots_axes_and_figures/subplots_demo.py]: `axs` here is an 2 by 2 array, so `ax` is also an array. For this to work, you need to flatten `axs`.

Comment (NelleV) [reply examples/subplots_axes_and_figures/subplots_demo.py]: I went ahead and fixed it. Hope that's fine!

**Pull request merged**